src/include/storage/metric/abstract_metric.h

```diff
@@ -0,0 +1,354 @@
+#pragma once
+
+#include <emmintrin.h>
+#include <atomic>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "catalog/catalog_defs.h"
+#include "storage/metric/abstract_raw_data.h"
+#include "storage/metric/metric_defs.h"
+
+namespace terrier {
+
+namespace transaction {
+class TransactionContext;
+}  // namespace transaction
+
+namespace storage::metric {
+/**
+ * @brief Interface representing a metric.
+ * A metric is defined to be some piece of logic that processes events generated
+ * by the database. @see StatsEventType for a list of available events.
+ * It is guaranteed that the appropriate callback method (identified by the
+ * naming pattern On[event name]) is invoked and the args filled out with
+ * relevant information. To enable safe and efficient collection of data,
+ * it is required that all data be collected be written to a RawData
+ * object, @see AbstractRawData.
+ *
+ * While you could write your own metric directly extending from this class,
+ * it is recommended that you use @see AbstractMetric class, which takes in
+ * an AbstractRawData class as a template argument and implements the tricky
+ * concurrent code for you.
+ *
+ * To write a new Metric, first write your own RawData class, extending from
+ * AbstractRawData, and extend from AbstractMetric with your RawData class as
+ * template argument. Then, override the event callbacks that you wish to know
+ * about. @see AbstractMetric on how to deal with concurrency.
+ */
+class Metric {
+ public:
+  virtual ~Metric() = default;
+
+  /**
+   * @param txn context of the transaction beginning
+   */
+  virtual void OnTransactionBegin(const transaction::TransactionContext *txn) {}
+
+  /**
+   * @param txn context of the transaction committing
+   * @param database_oid OID of the database where the txn happens.
+   */
+  virtual void OnTransactionCommit(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid) {}
+
+  /**
+   * @param txn context of the transaction aborting
+   * @param database_oid OID of the database where the txn happens.
+   */
+  virtual void OnTransactionAbort(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid) {}
+
+  /**
+   * @param txn context of the transaction performing read
+   * @param database_oid OID of the database that the tuple read happens
+   * @param namespace_oid OID of the namespace that the tuple read happens
+   * @param table_oid OID of the table that the tuple read happens
+   */
+  virtual void OnTupleRead(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid,
```

Comment:
Is it intentional that this method does not take the tuple (or a representation of it)?